cdap-security/src/main/java/co/cask/cdap/security/guice/SecureStoreModules.java

```diff
@@ -66,6 +70,26 @@ protected void configure() {
     };
   }
 
+  private static Object getStore(final CConfiguration cConf, Injector injector) {
+    if (AbstractSecureStore.isKMSBacked(cConf)) {
+      try {
+        if (AbstractSecureStore.isKMSCapable()) {
+          return injector.getInstance(Class.forName("co.cask.cdap.security.store.KMSSecureStore"));
+        } else {
+          LOG.warn("Could not find classes required for supporting KMS provider.");
+          // No Exception was thrown, that means the classes are available, load our provider
+          return injector.getInstance(DummyKMSStore.class);
+        }
+      } catch (ClassNotFoundException e) {
+        // KMSSecureStore could not be loaded
+        LOG.warn("Could not find classes required for supporting KMS based secure store.");
+        return injector.getInstance(DummyKMSStore.class);
+      }
+    } else {
+      return injector.getInstance(FileSecureStore.class);
+    }
+  }
+
   private static class SecureStoreProvider implements Provider<SecureStore> {
```

Comment:
You don't need two different providers. Just have one as:

```java
@Singleton
private static final class StoreProvider<T> implements Provider<T> {
  @Override
  public T get() {
    // You can move the static getStore method to this.
    return (T) store;
  }
}
```

Then in binding:

```java
bind(SecureStore.class).toProvider(new TypeLiteral<StoreProvider<SecureStore>>(){ });
bind(SecureStore.class).toProvider(new TypeLiteral<StoreProvider<SecureStoreManager>>(){ });
```